zone/task_manager.cpp

```diff
@@ -1639,18 +1553,262 @@ bool TaskManager::LoadClientState(Client *client, ClientTaskState *client_task_s
+		client_task_state->m_active_shared_task.task_id,
+		client_task_state->m_active_shared_task.slot
+	);
+	if (client_task_state->m_active_shared_task.task_id != TASKSLOTEMPTY) {
+		client_task_state->UnlockActivities(character_id, client_task_state->m_active_shared_task);
+	}
+
+	// quests (max 20 or 40 depending on client)
 	for (auto &active_quest : client_task_state->m_active_quests) {
 		if (active_quest.task_id != TASKSLOTEMPTY) {
 			client_task_state->UnlockActivities(character_id, active_quest);
 		}
 	}
 
-	LogTasks("[LoadClientState] for Character ID [{}] DONE!", character_id);
+	LogTasksDetail("[LoadClientState] for Character ID [{}] DONE!", character_id);
+	LogTasksDetail("---", character_id);
 
 	return true;
 }
+
+void TaskManager::SyncClientSharedTaskState(Client *c, ClientTaskState *cts)
+{
+	LogTasksDetail(
+		"[SyncClientSharedTaskState] Syncing client shared task state"
+	);
+
+	SyncClientSharedTaskWithPersistedState(c, cts);
+	SyncClientSharedTaskRemoveLocalIfNotExists(c, cts);
+}
+
+void TaskManager::SyncClientSharedTaskWithPersistedState(Client *c, ClientTaskState *cts)
+{
+	auto character_tasks = CharacterTasksRepository::GetWhere(
+		database,
+		fmt::format("charid = {} ORDER BY acceptedtime", c->CharacterID())
+	);
+
+	for (auto &character_task: character_tasks) {
+		if (character_task.type == TASK_TYPE_SHARED) {
+			auto st = SharedTaskMembersRepository::GetWhere(
+				database,
+				fmt::format(
+					"character_id = {}",
+					c->CharacterID()
+				)
+			);
+
+			if (!st.empty()) {
+				int64 shared_task_id = st[0].shared_task_id;
+				auto  activities     = SharedTaskActivityStateRepository::GetWhere(
+					database,
+					fmt::format(
+						"shared_task_id = {}",
+						shared_task_id
+					)
+				);
+
+				ClientTaskInformation *shared_task = nullptr;
+				shared_task = &cts->m_active_shared_task;
+
+				// has active shared task
+				if (shared_task) {
+
+					LogTasksDetail(
+						"[SyncClientSharedTaskWithPersistedState] Client [{}] has shared_task, sync with database",
+						c->GetCleanName()
+					);
+
+					bool      fell_behind_state = false;
+					for (auto &a: activities) {
+
+						LogTasksDetail(
+							"[LoadClientState] shared_task loop local [{}] shared [{}]",
+							shared_task->activity[a.activity_id].done_count,
+							a.done_count
+						);
+
+						// we're behind shared task state, update self
+						if (shared_task->activity[a.activity_id].done_count < a.done_count) {
+
+							// update done count
+							shared_task->activity[a.activity_id].done_count = a.done_count;
+
+							// activity state
+							shared_task->activity[a.activity_id].activity_state =
+								(a.completed_time > 0 ? ActivityCompleted : ActivityHidden);
+
+							// set flag to persist later
+							fell_behind_state = true;
+						}
+					}
+
+					// fell behind, force a save of client state
+					if (fell_behind_state) {
+						SaveClientState(c, cts);
+					}
+
+					c->SetSharedTaskId(shared_task_id);
```

Comment:
@Akkadius Does this need to be set to 0 in places where characters are removed from shared tasks? I didn't see any usage like that unless it's getting reset in another way

Reply:
Good point, zeroed out in loading client task state as well as when a client is removed from shared task